certora/specs/HashConsensus.spec

```diff
@@ -0,0 +1,742 @@
+    // keeps the current reference slot the same or
+    // increases it by at least the minimum of old and new frame sizes
+    uint256 slotsPerEpoch; uint256 secondsPerSlot; uint256 genesisTime;
+    slotsPerEpoch, secondsPerSlot, genesisTime = getChainConfig(e2);
+
+    assert refSlot2 >= refSlot1;  // fails
+
+    // assert (epochsPerFrame2 >= epochsPerFrame1) =>
+    //            ( (refSlot1 == refSlot2) || (refSlot2 >= refSlot1 + epochsPerFrame1 * slotsPerEpoch) );
+    
+    // assert (epochsPerFrame2 <= epochsPerFrame1) =>
+    //            ( (refSlot1 == refSlot2) || (refSlot2 >= refSlot1 + epochsPerFrame2 * slotsPerEpoch) );
+}
+
+
+//  4. setFastLaneLengthSlots() works as expected, setting it to zero disables the fast lane subset
+//     verify with getFrameConfig()
+// Status: Pass
+// https://prover.certora.com/output/80942/133f5763d704400b8bf17e08db8d01d6/?anonymousKey=c467fc94b9572fb48f8790f6c5ba5872f188ea1b
+rule setFastLaneLengthSlotsCorrectness() {
+    env e; 
+    setFastLaneLengthSlots(e,0);
+
+    uint256 initialEpoch; uint256 epochsPerFrame; uint256 fastLaneLengthSlots;
+    initialEpoch, epochsPerFrame, fastLaneLengthSlots = getFrameConfig(e);
+
+    assert fastLaneLengthSlots == 0;  // this assert passes: https://prover.certora.com/output/80942/133f5763d704400b8bf17e08db8d01d6/?anonymousKey=c467fc94b9572fb48f8790f6c5ba5872f188ea1b
+}
+
+
+//  5. addMember() - cannot add an existing member
+//  6. addMember() - cannot add an empty address as a member
+// Status: Pass
+// https://prover.certora.com/output/80942/d9d92700466f42b9b284bb7c0664314e/?anonymousKey=dfbcd43aa6f2a327b2fbe32e2ed4b51a06f52f3e
+rule addMemberRevertsCorrectly() {
+    env e; env e2;
+    address userA; uint256 quorum;
+    bool isUserAMember = getIsMember(e,userA);
+
+    addMember@withrevert(e, userA, quorum);                 // call addMember to add user userA
+    bool callReverted = lastReverted;
+
+    assert isUserAMember => callReverted;                   // cannot add existing member, rule 5
+    assert (userA == 0) => callReverted;                    // cannot add empty address, rule 6
+}
+
+
+//  7, addMember() - adding a member does not add/remove any other member
+// Status: Pass
+// https://prover.certora.com/output/80942/c6333d3aad5f47a19a3c91bb05f04e72/?anonymousKey=8ee21a783e8c2bf20dec20e40174f309a5958098
+rule addMemberADoesNotModifyMemberB() {
+    env e; env e2;
+    address userA; uint256 quorum; address userB;
+    bool isUserAMemberBefore = getIsMember(e,userA);
+    bool isUserBMemberBefore = getIsMember(e,userB);
+
+    addMember(e, userA, quorum);                            // call addMember to add user userA
+
+    bool isUserAMemberAfter = getIsMember(e2,userA);
+    bool isUserBMemberAfter = getIsMember(e2,userB);
+
+    require userA != userB;                                 // userA and userB are different
+
+    assert isUserBMemberBefore => isUserBMemberAfter;       // adding userA does not remove userB
```

Comment:
Maybe I didn't understand your intention, but it seems that fetching the member state of userA is not needed, and these two asserts could be combined into one with double implication (<=>).